src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -207,5 +240,41 @@ public Task UnloadAsync()
 
             return Task.CompletedTask;
         }
+
+        // true means the project system plans to call NominateProjectAsync in the future.
+        public bool HasPendingNomination()
+        {
+            return _pendingNomination;
+        }
+
+        // NuGet calls this method to wait project to nominate restoring
+        // If the project has no pending restore data, it will return a completed task.
+        // Otherwise a task which will be completed once the project norminate the next restore
+        // the task will be cancelled, if the project system decide it no longer need restore (for example: the restore state has no change)
+        // the task will be failed, if the project system runs into a problem, so it cannot get correct data to norminate a restore (DT build failed)
+        public Task WhenRestoreNominated()
+        {
+            if (_pendingNomination == false)
+            {
+                return Task.CompletedTask;
+            }
+            else
+            {
+                // task which will be completed once the project norminate the next restore (_norminateNextRestore == 1)
+                JoinableTask joinableTask = JoinableFactory.RunAsync(() =>
```

Comment:
No need do JTF work here. Our protocol is that NuGet never block UI thread waiting a normination, which we happen to lead into a dead lock.  (Also, the JTF work during waiting is always to Join the collection, and waiting on something async.  This code will hang the product.